### ISSUE TITLE
Feature request: "Run & Debug" feature without `launch.json`

### ISSUE THREAD

**Issue by tats-u** [fwcd/vscode-kotlin#128]
I don't know whether this issue should be here or in the debug adapter.

### Motivation

![image](https://github.com/fwcd/kotlin-language-server/assets/12870451/31c8956f-625a-4777-9722-ac428ce27155)

The Java extension of VS Code can run or debug (the `main` function of) `.java` files in the opened folder without `launch.json`.
This allows Java beginners to try their piece of code easily.

### Description

Related: #17

![image](https://github.com/fwcd/kotlin-language-server/assets/12870451/a99f534a-337e-46fa-aa92-1087ec6e05cf)

Currently, Kotlin files can't be run without `launch.json`. No "Run | Debug" in the top of `fun main`.

### Alternatives considered

`launch.json`: too complicated

**Comment by themkat** [comment]
That should already be available (ref: #92 ). My main guess is that it might not work on empty package names. Have you tried giving a package name? Looking for some output in the Kotlin-dropdown field in the Output menu might help. It is in the lower right corner where you find the terminal etc.:
<img width="964" alt="image" src="https://github.com/fwcd/vscode-kotlin/assets/5732795/aa301614-2ca0-471e-bc65-3960e294ed61">

**Comment by tats-u** [comment]
@themkat Just opened a folder without `pom.xml` or `build.gradle(.kts)`.
I'll change the title.

Does #92 make it possible to start debugging without `launch.json`?

**Comment by themkat** [comment]
> Did #92 make it possible to start debugging without `launch.json`?

Yes. It gave you the Run/Debug code lenses like you see in the Java example you provided. 

When you mention it, the debug adapter isn't really working without Maven or Gradle. So even if you got the Run/Debug code lenses working, I guess you would not be able to run it. There is an open issue on this for the debug adapter: https://github.com/fwcd/kotlin-debug-adapter/issues/62

**Comment by tats-u** [comment]
@themkat 
Can you start debugging only with pressing <kbd>F5</kbd> (not CodeLens action) in the main source file?
I can't prepare a project that can run in VS Code. I created one with IDEA with Kotlin + Gradle but didn't work with VS Code.

**Comment by tats-u** [comment]
https://github.com/ktorio/ktor-samples/tree/main/chat

I cloned and opened this directory but CodeLens around any of `main`s there.

**Comment by themkat** [comment]
> @themkat Can you start debugging only with pressing F5 (not CodeLens action) in the main source file? I can't prepare a project that can run in VS Code. I created one with IDEA with Kotlin + Gradle but didn't work with VS Code. The Gradle extension didn't work well for me, either.

No. That would off course require a launch.json. How else would it know what to run? I must have misunderstood you at first, because I thought it was about missing code lens.

---

> https://github.com/ktorio/ktor-samples/tree/main/chat
> 
> I cloned and opened this directory but CodeLens around any of mains there.

That project might sadly not work, even for some basic operations. The language server is not really good with the Kotlin multiplatform structures and stuff. The support for it, just fetching dependencies is very experimental. https://github.com/fwcd/kotlin-language-server/pull/377

Tried building (to fetch dependencies), but the file is full of resolving issues. Not able to debug, but I get the run/debug-codelens in a weird place (above the comment):
<img width="1333" alt="image" src="https://github.com/fwcd/vscode-kotlin/assets/5732795/1635cf48-6ca9-4945-8a74-8e8d21f4ce09">

The code lens works, but there seems to be other issues. Probably due to Kotlin multiplatform structures that the debug adapter is not used to looking for. Sadly I don't know enough about Kotlin multiplatform to find a fix. Might just be as simple as fixing some directory lookup here (in the debug adapter):
https://github.com/fwcd/kotlin-debug-adapter/blob/main/adapter/src/main/kotlin/org/javacs/ktda/classpath/ProjectClassesResolver.kt
(from what I can see, Multiplatform projects have directories like `backend`, `js` etc. inside of `build/classes`, instead of a more flat structure)

---

For anyone not getting the Run/Debug code lens:
Make sure that the debug adapter is enabled in the VSCode Kotlin extension settings: 
<img width="916" alt="image" src="https://github.com/fwcd/vscode-kotlin/assets/5732795/f0d42023-7089-409b-be51-172b2af9f8cf">
If it does not show up at once, restart your editor. 

From a stupid Spring Boot project I just threw together:
<img width="737" alt="image" src="https://github.com/fwcd/vscode-kotlin/assets/5732795/4d87ea31-19f0-4d84-a0c4-2b9acd2a5940">

Beyond these basics, I really don't know much about VSCode. Even if I have contributed to this extension, I NEVER use it beyond simple experiments. Emacs is the only true editor 😛
